Scripts for events pipeline

ajaits opened the pull request:
Scripts for events pipeline to run a sequence of operations:
- download source data
- generate geoTif from EarthEngine for a dataset and bands
- generate CSV with S2 cells for places from geoTif or source  files
- process CSV into events and related SVObs per-event. per-place

jehangiramjad commented on scripts/earthengine/utils.py:
```diff
@@ -649,3 +435,61 @@ def str_format_float(data: float, precision_digits: int = 6):
+    re_pat = r'P?(?P<delta>[+-]?[0-9]+)(?P<unit>[A-Z])'
+    m = re.search(re_pat, time_period.upper())
+    if m:
+        m_dict = m.groupdict()
+        delta = int(m_dict.get('delta', '0'))
+        unit = m_dict.get('unit', 'M')
+        period_dict = {'D': 'days', 'M': 'months', 'Y': 'years'}
+        period = period_dict.get(unit, 'day')
```
nit: would be great to have a couple of comments explaining what's going on here.

ajaits replied:
Added more comments.

jehangiramjad commented on scripts/earthengine/utils.py:
```diff
@@ -649,3 +435,61 @@ def str_format_float(data: float, precision_digits: int = 6):
+                           time_period: str,
+                           date_format: str = '%Y-%m-%d') -> str:
+    '''Returns the date string advanced by the time period.'''
+    next_date = ''
```
nit: looks like `next_date` variable is only ever used to store the empty string. So perhaps wherever this function is returning `next_date`, it can just return the empty string ('') instead of `next_date`?

ajaits replied:
removed next_date

jehangiramjad commented on scripts/earthengine/utils.py:
```diff
@@ -649,3 +435,61 @@ def str_format_float(data: float, precision_digits: int = 6):
+
+
+def date_format_by_time_period(date_str: str, time_period: str) -> str:
+    '''Returns date in the format of the timeperiod.
```
nit: the function comment (docstring) formatting seems off?

jehangiramjad commented on scripts/earthengine/earthengine_image.py:
```diff
@@ -634,11 +612,18 @@ def ee_process(config) -> list:
+    cur_date = utils.date_today()
+    counters = Counters()
+    # Get images by count or until yesterday
+    while (config['ee_image_count'] > 0) and (config.get('start_date', '') <
```
nit: instead of `config['ee_image_count'] > 0` I think you can simply do: `config['ee_image_count']`

ajaits replied:
Done

jehangiramjad commented on scripts/earthengine/earthengine_image.py:
```diff
@@ -649,29 +634,54 @@ def ee_process(config) -> list:
+        if task.active():
+            pending_tasks.add(task)
+            logging.info(f'Waiting for task: {task}')
+            time.sleep(10)
```
nit: ideally, have the timeout (sleep) seconds be a constant somewhere so its easy to change?

ajaits replied:
Done

jehangiramjad commented on scripts/earthengine/earthengine_image.py:
```diff
@@ -649,29 +634,54 @@ def ee_process(config) -> list:
+
+
+def get_date_from_filename(filename: str, time_period: str = '') -> str:
+    '''Returns the date from the filename.'''
+    matches = re.search(r'from_(?P<date>[0-9]{4}[0-9-]{2}[0-9-]{0,2})',
+                        filename)
+    if matches:
+        return utils.date_format_by_time_period(
+            matches.groupdict().get('date', ''), time_period)
+    return ''
```
nit: this feels like a lot of magic :) so a couple of comments would help.

ajaits replied:
Added some comments.

jehangiramjad commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+          config: dictionary of configuration settings for this stage.
+          counters: global counters for the task.
+       Returns:
+         list if output files that are passed on to the next stage as inputs.
```
nit: list *of*

ajaits replied:
Done

jehangiramjad commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+                    f'Setting start_date to start of year: {start_date}')
+            if start_date:
+                self.set_config('start_date', start_date)
+        self.set_config('year', start_date[:4])
+        self.set_config('month', start_date[5:7])
+        self.set_config('day', start_date[8:10])
```
nit: might be preferable to split by "-" and then get index 0, 1 and 2?

ajaits replied:
Done

jehangiramjad commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+        self.set_config('day', start_date[8:10])
+
+
+class EarthEngineRunner(StageRunner):
```
nit: perhaps in a follow-up PR, should each of these specific Runners be in their own file?

pradh replied:
+1

ajaits replied:
Split runners into separate files pipeline_stage_*.py
test is a single file though for now, events_pipeline_test.py tests multiple stages.

jehangiramjad commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script to generate events from source data.
+It is a pipeline with the following steps:
+  1. get_source_data: Download data from source, incrementally with any source specific processing
+  2. generate_gridded_csv: Process source data into CSV with S2 cells.
+  3. generate_events: Generate events and related stats from the csv data.
+"""
```
nit: since this is the most crucial part of the processing, I'd love to have more comments. For example:

is there an order to the processing? is it done in parallel? what if there is a failure in some stage of the pipeline (does it need to be re-run or can it just resume from where it left of)? etc.

pradh replied:
+1

ajaits replied:
Added more comments on these points.

jehangiramjad reviewed (approved):
Left lots of nitty comments which can also be addressed later. This was a lot of code for reviewing so I may have missed some things. Over all, it would be great to provide comprehensive details about the various failures modes of the events_pipeline processors :)

pradh commented on scripts/earthengine/README.md:
```diff
@@ -139,6 +139,19 @@ python3 process_events.py \
+downloading data from URLs or extracting geoTiff from EarthEngine and
+generate events from the data.
+
+It uses a config with settings for each stage and can resume processing
```
Could we document the config?

If a dev wants to use the events pipeline for a new type of event, what would it entail?

ajaits replied:
Added some brief overview of the config. I'll add more examples and instructions in a following PR.

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -1034,29 +1136,239 @@ def read_active_events(self, filename: str):
+                        aggregate_dict(
+                            pvs, place_date_pvs[key],
+                            self._config.get('property_config_per_date',
+                                             {'aggregate': 'sum'}))
```
Could aggregation be in a helper func?

ajaits replied:
Done

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -1034,29 +1136,239 @@ def read_active_events(self, filename: str):
+                for prop in event_props:
+                    if prop in pvs:
+                        row_dict[prop] = pvs[prop]
+                writer.writerow(_format_property_values(row_dict, digits=3))
```
Could CSV writing be in a helper func?

ajaits replied:
Done

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -1034,29 +1136,239 @@ def read_active_events(self, filename: str):
+        )
+        with file_util.FileIO(output_tmcf, 'w') as tmcf_file:
+            tmcf_file.write('\n\n'.join(tmcf_nodes))
+        output_files.append(output_tmcf)
```
Could TMCF writing be in a helper func?

ajaits replied:
Done

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -1034,29 +1136,239 @@ def read_active_events(self, filename: str):
         logging.info(
             f'Loaded {len(active_events)} events from file: {filename}')
 
+    def write_events_place_svobs(
```
This is an incredibly long function :). I would recommend adding more comments on the high level flow

ajaits replied:
Added a comment and broke it down into smaller functions.

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -1034,29 +1136,239 @@ def read_active_events(self, filename: str):
+            'EventId': {
+                'aggregate': 'set'
+            }
+        }
```
nit: Can these constants be hoisted out to make the function logic more accessible?

ajaits replied:
Done

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -531,24 +563,67 @@ def deactivate_old_events(self, event_ids: list, date: str) -> list:
-                                   event_id)
+
+    def merge_events(self, event_ids: list) -> GeoEvent:
+        '''returns root_event, which is the earliest event in the events_list and 
```
What is "root_event"?

ajaits replied:
Added some more comments. Pls let me know of any alternate phrase that is easier to follow :)

pradh commented on scripts/earthengine/process_events.py:
```diff
@@ -159,10 +162,19 @@
     'polygon_simplification_factor': 0.1,
 
+    # Output svobs per place
+    'output_place_svobs': True,
```
For my understanding, where are the keys described?

Also, wonder if we can use separate files for config constants? (later PR is fine)

ajaits replied:
Moved the config to event_config.py with more comments describing each setting.
That is the doc for config for now :)

pradh commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+
+_DEBUG = False
+
+# Default configuration settings.
```
Can we keep configs in a separate file?

ajaits replied:
Moved it to event_config.py

pradh commented on scripts/earthengine/events_pipeline.py:
```diff
@@ -0,0 +1,818 @@
+        )
+        return []
+
+    def setUp(self,
```
nit: why setUp and not set_up :)

ajaits replied:
was using convention similar to unittest.textCase.setUp()
changed to set_up()

pradh commented on scripts/earthengine/test_data/sample_fires_event_pipeline_config.py:
```diff
@@ -0,0 +1,111 @@
+{
```
We might want to document this (later PR).  I'm not sure if someone other than Ajai can setup event configs :)

ajaits replied:
I'll add a separate config.py with details of each stage config.

pradh reviewed (approved):
Left some comments.  Please feel free to defer as necessary.

Generally I would recommend breaking up files, keeping maps separate, and modularizing functions... in future PRs :)

ajaits commented on scripts/earthengine/utils.py:
```diff
@@ -142,7 +96,17 @@ def is_s2_cell_id(dcid: str) -> bool:
+    Args:
+      lat: Latitude in degrees
+      lng: Longitude in degrees
+      level: desired S2 level for cell id in the range 0-30
```
S2 levels supported are 0:30.
https://s2geometry.io/resources/s2cell_statistics

ajaits commented on scripts/earthengine/utils.py:
```diff
@@ -386,7 +355,7 @@ def place_distance(place1: str, place2: str) -> float:
 def place_area(place: str) -> float:
     '''Returns the area for the place.'''
```
Added sqkm.

ajaits commented:
> Left lots of nitty comments which can also be addressed later. This was a lot of code for reviewing so I may have missed some things. Over all, it would be great to provide comprehensive details about the various failures modes of the events_pipeline processors :)

Thanks Jehangir. I have addressed the comments and added more descriptions.

I can add more on failure modes in a followup README

ajaits commented:
> Left some comments. Please feel free to defer as necessary.
> 
> Generally I would recommend breaking up files, keeping maps separate, and modularizing functions... in future PRs :)

Thanks Prashanth. I have addressed the comments and also broken up the pipeline files.

PTAL.

pradh reviewed (approved):
Thanks for the refactors Ajai!
